File: ChatSDKDemo/Assets/ChatSDK/Plugins/iOS/HyphenateWrapper/Wrapper/EMConversationWrapper.m
//
//  EMConversationWrapper.m
//  HyphenateWrapper
//
//  Created by 杜洁鹏 on 2021/6/15.
//

#import "EMConversationWrapper.h"
#import "EMConversation+Unity.h"
#import "Transfrom.h"
#import "EMChatMessage+Unity.h"

@implementation EMConversationWrapper
- (void)getConversationWithParam:(NSDictionary *)param
                      completion:(void(^)(EMConversation *conversation))aCompletion
{
    __weak NSString *conversationId = param[@"convId"];
    EMConversationType type = [EMConversation typeFromInt:[param[@"convType"] intValue]];
    EMConversation *conversation = [EMClient.sharedClient.chatManager getConversation:conversationId
                                                                                 type:type
                                                                     createIfNotExist:YES];
    if (aCompletion) {
        aCompletion(conversation);
    }
}

#pragma mark - Actions
- (NSDictionary *)getUnreadMsgCount:(NSDictionary *)param
{
    __block NSDictionary *ret = nil;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation) {
        ret = @{@"count": @(conversation.unreadMessagesCount)};
    }];
    
    return ret;
}

- (NSDictionary *)getLatestMessage:(NSDictionary *)param
{
    __block NSDictionary *ret = nil;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        EMChatMessage *msg = conversation.latestMessage;
        ret = msg.toJson;
    }];
    return ret;
}

- (NSDictionary *)getLatestMessageFromOthers:(NSDictionary *)param
{
    __block NSDictionary *ret = nil;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        EMChatMessage *msg =conversation.lastReceivedMessage;
        ret = msg.toJson;
    }];
    return ret;
}

- (void)markMessageAsRead:(NSDictionary *)param
{
    __block NSString *msgId = param[@"msgId"];
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {  
        EMError *error = nil;
        [conversation markMessageAsReadWithId:msgId error:&error];
    }];
}

- (void)syncConversationExt:(NSDictionary *)param
{
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        NSDictionary *ext = [Transfrom NSStringToJsonObject: param[@"ext"]];
        conversation.ext = ext;
    }];
}

- (NSDictionary *)conversationExt:(NSDictionary *)param {
    __block NSDictionary *ret = nil;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        ret = conversation.ext;
    }];
    return ret;
}

- (void)markAllMessagesAsRead:(NSDictionary *)param
{
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation) {
        [conversation markAllMessagesAsRead:nil];
    }];
}

- (NSDictionary *)insertMessage:(NSDictionary *)param
{
    __block NSDictionary *ret = nil;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        NSDictionary *msgDict = [Transfrom NSStringToJsonObject:param[@"msg"]];
        EMChatMessage *msg = [EMChatMessage fromJson:msgDict];
        
        EMError *error = nil;
        [conversation insertMessage:msg error:&error];
        ret = @{@"ret": @(error == nil)};
    }];
    return ret;
}

- (NSDictionary *)appendMessage:(NSDictionary *)param
{
    __block NSDictionary *ret = nil;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        NSDictionary *msgDict = [Transfrom NSStringToJsonObject:param[@"msg"]];
        EMChatMessage *msg = [EMChatMessage fromJson:msgDict];
        
        EMError *error = nil;
        [conversation appendMessage:msg error:&error];
        ret = @{@"ret": @(error == nil)};
    }];
    return ret;
}

- (NSDictionary *)messageCount:(NSDictionary *)param {
    __block NSDictionary *ret = nil;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation) {
        ret = @{@"count": @(conversation.messagesCount)};
    }];
    
    return ret;
}

- (NSDictionary *)updateConversationMessage:(NSDictionary *)param
{
    __block NSDictionary *ret = nil;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        NSDictionary *msgDict = [Transfrom NSStringToJsonObject:param[@"msg"]];
        EMChatMessage *msg = [EMChatMessage fromJson:msgDict];
        
        EMError *error = nil;
        [conversation updateMessageChange:msg error:&error];
        ret = @{@"ret": @(error == nil)};
    }];
    return ret;
}

- (NSDictionary *)removeMessage:(NSDictionary *)param
{
    __block NSDictionary *ret = nil;
    __block NSString *msgId = param[@"msgId"];
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        EMError *error = nil;
        [conversation deleteMessageWithId:msgId error:&error];
        ret = @{@"ret": @(error == nil)};
    }];
    return ret;
}

- (NSDictionary *)clearAllMessages:(NSDictionary *)param
{
    __block NSDictionary *ret = nil;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        EMError *error = nil;
        [conversation deleteAllMessages:&error];
        ret = @{@"ret": @(error == nil)};
    }];
    return ret;
}

#pragma mark - load messages
- (NSDictionary *)loadMsgWithId:(NSDictionary *)param
{
    __block NSString *msgId = param[@"msgId"];
    __block NSDictionary *ret = nil;
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        EMError *error = nil;
        EMChatMessage *msg = [conversation loadMessageWithId:msgId error:&error];
        ret = msg.toJson;
    }];
    return ret;
}

- (void)loadMsgWithMsgType:(NSDictionary *)param callbackId:(NSString *)callbackId
{
    __block NSString *callId = callbackId;
    EMMessageBodyType type = [EMMessageBody typeFromString:param[@"type"]];
    long long timeStamp = [param[@"timeStamp"] longLongValue];
    int count = [param[@"count"] intValue];
    NSString *sender = param[@"sender"];
    EMMessageSearchDirection direction = [self searchDirectionFromString:param[@"direction"]];
    
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        if (conversation) {
            [conversation loadMessagesWithType:type
                                     timestamp:timeStamp
                                         count:count
                                      fromUser:sender
                               searchDirection:direction
                                    completion:^(NSArray *aMessages, EMError *aError)
             {
                if (aError) {
                    [self onError:callId error:aError];
                }else {
                    NSMutableArray *ary = [NSMutableArray array];
                    for (EMChatMessage *msg in aMessages) {
                        [ary addObject:[Transfrom NSStringFromJsonObject:[msg toJson]]];
                    }
                    [self onSuccess:@"List<EMMessage>" callbackId:callId userInfo:[Transfrom NSStringFromJsonObject:ary]];
                }
            }];
        }
    }];

}

- (void)loadMsgWithStartId:(NSDictionary *)param callbackId:(NSString *)callbackId
{
    __block NSString *startId = param[@"startId"];
    __block NSString *callId = callbackId;
    int count = [param[@"count"] intValue];
    EMMessageSearchDirection direction = [self searchDirectionFromString:param[@"direction"]];

    [self getConversationWithParam:param completion:^(EMConversation *conversation)
    {
        if (conversation) {
            [conversation loadMessagesStartFromId:startId
                                            count:count
                                  searchDirection:direction
                                       completion:^(NSArray *aMessages, EMError *aError)
             {
                if (aError) {
                    [self onError:callId error:aError];
                }else {
                    NSMutableArray *ary = [NSMutableArray array];
                    for (EMChatMessage *msg in aMessages) {
                        [ary addObject:[Transfrom NSStringFromJsonObject:[msg toJson]]];
                    }
                    [self onSuccess:@"List<EMMessage>" callbackId:callId userInfo:[Transfrom NSStringFromJsonObject:ary]];
                }
            }];
        }
    }];
}

- (void)loadMsgWithKeywords:(NSDictionary *)param callbackId:(NSString *)callbackId
{
    __block  NSString * callId = callbackId;
    __block  NSString * keywords = param[@"keywords"];
    long long timestamp = [param[@"timestamp"] longLongValue];
    int count = [param[@"count"] intValue];
    __block  NSString *sender = param[@"sender"];
    EMMessageSearchDirection direction = [self searchDirectionFromString:param[@"direction"]];
    
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        if (conversation) {
            [conversation loadMessagesWithKeyword:keywords
                                        timestamp:timestamp
                                            count:count
                                         fromUser:sender
                                  searchDirection:direction
                                       completion:^(NSArray *aMessages, EMError *aError)
             {
                if (aError) {
                    [self onError:callId error:aError];
                }else {
                    NSMutableArray *ary = [NSMutableArray array];
                    for (EMChatMessage *msg in aMessages) {
                        [ary addObject:[Transfrom NSStringFromJsonObject:[msg toJson]]];
                    }
                    [self onSuccess:@"List<EMMessage>" callbackId:callId userInfo:[Transfrom NSStringFromJsonObject:ary]];
                }
            }];
        }
    }];
}

- (void)loadMsgWithTime:(NSDictionary *)param callbackId:(NSString *)callbackId
{
    __block NSString * callId = callbackId;
    long long startTime = [param[@"startTime"] longLongValue];
    long long entTime = [param[@"endTime"] longLongValue];
    int count = [param[@"count"] intValue];
    
    [self getConversationWithParam:param
                        completion:^(EMConversation *conversation)
     {
        if (conversation) {
            [conversation loadMessagesFrom:startTime
                                        to:entTime
                                     count:count
                                completion:^(NSArray *aMessages, EMError *aError)
             {
                if (aError) {
                    [self onError:callId error:aError];
                }else {
                    NSMutableArray *ary = [NSMutableArray array];
                    for (EMChatMessage *msg in aMessages) {
                        [ary addObject:[Transfrom NSStringFromJsonObject:[msg toJson]]];
                    }
                    [self onSuccess:@"List<EMMessage>" callbackId:callId userInfo:[Transfrom NSStringFromJsonObject:ary]];
                }
            }];
        }
    }];
}



- (EMMessageSearchDirection)searchDirectionFromString:(NSString *)aDirection {
    return [aDirection isEqualToString:@"up"] ? EMMessageSearchDirectionUp : EMMessageSearchDirectionDown;
}
@end

File: Unity_SDK/android_build/unity_chat_sdk/src/main/java/com/hyphenate/unity_chat_sdk/EMConversationWrapper.java
package com.hyphenate.unity_chat_sdk;

import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMConversation;
import com.hyphenate.chat.EMMessage;
import com.hyphenate.unity_chat_sdk.helper.EMConversationHelper;
import com.hyphenate.unity_chat_sdk.helper.EMMessageHelper;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

public class EMConversationWrapper extends EMWrapper{
    static public EMConversationWrapper wrapper() {
        return new EMConversationWrapper();
    }

    private boolean appendMessage(String conversationId, int conversationType, String messageString) throws JSONException {

        if (conversationId == null || conversationId.length() == 0) {
            return false;
        }

        EMConversation conversation = getConversation(conversationId, conversationType);
        EMMessage message = EMMessageHelper.fromJson(new JSONObject(messageString));
        conversation.appendMessage(message);
        return true;
    }

    private boolean deleteAllMessages(String conversationId, int conversationType) throws JSONException {
        EMConversation conversation = getConversation(conversationId, conversationType);
        conversation.clearAllMessages();
        return true;
    }

    private boolean removeMessage(String conversationId, int conversationType, String messageId){
        if (conversationId == null || conversationId.length() == 0) {
            return false;
        }
        if (messageId == null || messageId.length() == 0) {
            return false;
        }
        EMConversation conversation = getConversation(conversationId, conversationType);
        conversation.removeMessage(messageId);
        return true;
    }

    private String getExt(String conversationId, int conversationType) {
        if (conversationId == null || conversationId.length() == 0) {
            return null;
        }
        EMConversation conversation = getConversation(conversationId, conversationType);
        if (conversation == null) return null;
        return conversation.getExtField();
    }

    private boolean insertMessage(String conversationId, int conversationType, String messageString) throws JSONException {
        if (conversationId == null || conversationId.length() == 0) {
            return false;
        }

        if (messageString == null || messageString.length() == 0) {
            return false;
        }

        EMConversation conversation = getConversation(conversationId, conversationType);
        EMMessage message = EMMessageHelper.fromJson(new JSONObject(messageString));
        conversation.insertMessage(message);
        return true;
    }

    private String lastMessage(String conversationId, int conversationType){

        if (conversationId == null || conversationId.length() == 0) {
            return null;
        }

        String ret = null;

        EMConversation conversation = getConversation(conversationId, conversationType);
        EMMessage msg = conversation.getLastMessage();
        if (msg != null) {
            try {
                ret = EMMessageHelper.toJson(msg).toString();
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return ret;
    }

    private String lastReceivedMessage(String conversationId, int conversationType){

        if (conversationId == null || conversationId.length() == 0) {
            return null;
        }

        String ret = null;
        EMConversation conversation = getConversation(conversationId, conversationType);
        EMMessage msg = conversation.getLatestMessageFromOthers();
        if (msg != null) {
            try {
                ret = EMMessageHelper.toJson(msg).toString();
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return ret;
    }

    private String loadMsgWithId(String conversationId, int conversationType, String msgId) {
        if (conversationId == null || conversationId.length() == 0) {
            return null;
        }

        if (msgId == null || msgId.length() == 0) {
            return null;
        }

        String ret = null;
        EMMessage msg = EMClient.getInstance().chatManager().getMessage(msgId);
        if (msg != null) {
            try {
                ret = EMMessageHelper.toJson(msg).toString();
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return ret;
    }

    private void loadMsgWithStartId(String conversationId, int conversationType, String startId, int count, String directionString, String callbackId) {
        if (conversationId == null || conversationId.length() == 0) {
            return;
        }
        String ret = null;
        EMConversation conversation = getConversation(conversationId, conversationType);
        List<EMMessage> msgList = conversation.loadMoreMsgFromDB(startId, count);
        JSONArray jsonArray = new JSONArray();
        try {
            for (EMMessage msg: msgList) {
                jsonArray.put(EMMessageHelper.toJson(msg).toString());
            }
            ret = jsonArray.toString();
        }  catch (JSONException e) {
            e.printStackTrace();
        }
        onSuccess("List<EMMessage>", callbackId, ret);
    }

    private void loadMsgWithKeywords(String conversationId, int conversationType, String keywords, String sender, long timestamp, int count, String directionString, String callbackId) {
        if (conversationId == null || conversationId.length() == 0) {
            return;
        }
        String ret = null;
        EMConversation conversation = getConversation(conversationId, conversationType);
        EMConversation.EMSearchDirection direction = searchDirectionFromString(directionString);
        List<EMMessage> msgList = conversation.searchMsgFromDB(keywords, timestamp, count, sender, direction);
        JSONArray jsonArray = new JSONArray();
        try {
            for (EMMessage msg: msgList) {
                jsonArray.put(EMMessageHelper.toJson(msg).toString());
            }
            ret = jsonArray.toString();
        }  catch (JSONException e) {
            e.printStackTrace();
        }

        onSuccess("List<EMMessage>", callbackId, ret);
    }

    private void loadMsgWithMsgType(String conversationId, int conversationType, String typeString, String sender, long timestamp, int count, String directionString, String callbackId) {
        if (conversationId == null || conversationId.length() == 0) {
            return;
        }
        String ret = null;
        EMConversation conversation = getConversation(conversationId, conversationType);
        EMConversation.EMSearchDirection direction = searchDirectionFromString(directionString);
        EMMessage.Type type = EMMessage.Type.TXT;
        switch (typeString) {
            case "txt" : type = EMMessage.Type.TXT; break;
            case "loc" : type = EMMessage.Type.LOCATION; break;
            case "cmd" : type = EMMessage.Type.CMD; break;
            case "custom" : type = EMMessage.Type.CUSTOM; break;
            case "file" : type = EMMessage.Type.FILE; break;
            case "img" : type = EMMessage.Type.IMAGE; break;
            case "video" : type = EMMessage.Type.VIDEO; break;
            case "voice" : type = EMMessage.Type.VOICE; break;
        }
        List<EMMessage> msgList = conversation.searchMsgFromDB(type, timestamp, count, sender, direction);
        JSONArray jsonArray = new JSONArray();
        try {
            for (EMMessage msg: msgList) {
                jsonArray.put(EMMessageHelper.toJson(msg).toString());
            }
            ret = jsonArray.toString();
        }  catch (JSONException e) {
            e.printStackTrace();
        }
        onSuccess("List<EMMessage>", callbackId, ret);
    }

    private void loadMsgWithTime(String conversationId, int conversationType, long startTime, long endTime, int count, String callbackId ) throws JSONException {
        if (conversationId == null || conversationId.length() == 0) {
            return;
        }
        String ret = null;
        EMConversation conversation = getConversation(conversationId, conversationType);
        List<EMMessage> msgList = conversation.searchMsgFromDB(startTime, endTime, count);
        JSONArray jsonArray = new JSONArray();
        try {
            for (EMMessage msg: msgList) {
                jsonArray.put(EMMessageHelper.toJson(msg).toString());
            }
            ret = jsonArray.toString();
        }  catch (JSONException e) {
            e.printStackTrace();
        }
        onSuccess("List<EMMessage>", callbackId, ret);
    }

    private void markAllMessagesAsRead(String conversationId, int conversationType) throws JSONException {
        if (conversationId == null || conversationId.length() == 0) {
            return;
        }
        EMConversation conversation = getConversation(conversationId, conversationType);
        conversation.markAllMessagesAsRead();
    }

    private void markMessageAsRead(String conversationId, int conversationType, String messageId) throws JSONException {
        if (conversationId == null || conversationId.length() == 0 || messageId == null || messageId.length() == 0) {
            return;
        }
        EMConversation conversation = getConversation(conversationId, conversationType);
        conversation.markMessageAsRead(messageId);
    }

    private void SetExt(String conversationId, int conversationType, String ext) {
        if (conversationId == null || conversationId.length() == 0) {
            return;
        }
        EMConversation conversation = getConversation(conversationId, conversationType);
        conversation.setExtField(ext);
    }

    private int unreadCount(String conversationId, int conversationType) {
        if (conversationId == null || conversationId.length() == 0) {
            return 0;
        }
        EMConversation conversation = getConversation(conversationId, conversationType);
        return conversation.getUnreadMsgCount();
    }

    private int messageCount(String conversationId, int conversationType) {
        if (conversationId == null || conversationId.length() == 0) {
            return 0;
        }
        EMConversation conversation = getConversation(conversationId, conversationType);
        return conversation.getAllMsgCount();
    }

    private boolean updateConversationMessage(String conversationId, int conversationType, String messageString) throws JSONException {
        if (conversationId == null || conversationId.length() == 0 || messageString == null || messageString.length() == 0) {
            return false;
        }
        EMConversation conversation = getConversation(conversationId, conversationType);
        EMMessage message = EMMessageHelper.fromJson(new JSONObject(messageString));
        conversation.updateMessage(message);
        return true;
    }

    private EMConversation getConversation(String conversationId, int conType ){

        EMConversation.EMConversationType type = EMConversationHelper.typeFromInt(conType);
        EMConversation conversation = EMClient.getInstance().chatManager().getConversation(conversationId, type, true);

        return conversation;
    }

    private EMConversation.EMSearchDirection searchDirectionFromString(String direction) {
        return direction.equals("up")  ? EMConversation.EMSearchDirection.UP : EMConversation.EMSearchDirection.DOWN;
    }
}
